File: library/src/main/java/com/getstream/sdk/chat/utils/State.java
package com.getstream.sdk.chat.utils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class State<T> {

    @NonNull
    public final T data;

    @NonNull
    public final Throwable error;

    public State(T data) {
        this.data = data;
        this.error = null;
    }

    public State(@Nullable Throwable error) {
        this.error = error;
        this.data = null;
    }

    public boolean success() {
        return data != null;
    }

    public boolean error() {
        return error != null;
    }
}

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/State.kt
package io.getstream.chat.android.client.socket

import androidx.annotation.VisibleForTesting
import io.getstream.chat.android.client.clientstate.DisconnectCause
import io.getstream.chat.android.client.errors.ChatNetworkError
import io.getstream.chat.android.client.events.ConnectedEvent

/**
 * State of the socket connection.
 */
@VisibleForTesting
internal sealed class State {

    /**
     * State of socket when connection is being established.
     */
    data class Connecting(val session: Session) : State()

    /**
     * State of socket when the connection is established.
     */
    data class Connected(val event: ConnectedEvent?, val session: Session) : State()

    /**
     * State of socket when the connection is permanently disabled.
     */
    class DisconnectedPermanently(val error: ChatNetworkError?) : State()

    /**
     * State of socket when connection is being disconnecting.
     */
    data class Disconnecting(val disconnectCause: DisconnectCause) : State()

    /**
     * State of socket when connection is disconnected.
     * The connection maybe established again based on [disconnectCause].
     */
    data class Disconnected(val disconnectCause: DisconnectCause) : State()

    /**
     * State of socket after it is destroyed and won't be reconnected.
     */
    object Destroyed : State()

    /**
     * Get connection id of this connection.
     */
    internal fun connectionIdOrError(): String = when (this) {
        is Connected -> event?.connectionId ?: error("This state doesn't contain connectionId")
        else -> error("This state doesn't contain connectionId")
    }
}